plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -848,6 +857,280 @@ compileExpr e = withContextM 2 (sdToTxt $ "Compiling expr:" GHC.<+> GHC.ppr e) $
+    Maybe (m (PIRTerm uni fun))
+isKnownPred ty v@(GHC.idDetails -> GHC.DataConWorkId _) args =
+    if GHC.isPredTy ty
+        then do
+            (tc, targs) <- GHC.splitTyConApp_maybe ty
+            cls <- GHC.tyConClass_maybe tc
+            f <- Map.lookup (splitGhcName (GHC.tyConName tc)) knownPreds
+            f targs (cls, v, args)
+        else Nothing
+isKnownPred _ _ _ = Nothing
+
+-- | Like `knownApps` but for class dictionaries.
+knownPreds ::
+    CompilingDefault uni fun m ann =>
+    Map (Maybe String, String)
+        ([GHC.Type] -> (GHC.Class, GHC.Var, [GHC.CoreExpr]) -> Maybe (m (PIRTerm uni fun)))
+knownPreds =
+    Map.fromListWithKey (\n -> error ("knownApps: key defined more than once: " <> show n))
+        . fmap (first splitThName)
+        $ [ ( ''GHC.Classes.Eq
+            , \case
+                -- `Eq Integer`
+                [ty] | ty `GHC.eqType` GHC.integerTy -> \(cls, v, args) ->
+                    Just $ compileDict cls v args Dictionary.Integer.selectorsEq
+                -- `Eq [a]`
+                [GHC.splitTyConApp_maybe -> Just (tc, [_targ])] | tc == GHC.listTyCon ->
+                    \(cls, v, args) ->
+                        Just $ compileDict cls v args Dictionary.List.selectorsEq
+                _ -> const Nothing
+            ),
+            ( ''GHC.Classes.Ord
+            , \case
+                -- `Ord Integer`
+                [ty] | ty `GHC.eqType` GHC.integerTy -> \(cls, v, args) ->
+                    Just $ compileDict cls v args Dictionary.Integer.selectorsOrd
+                -- `Ord [a]`
+                [GHC.splitTyConApp_maybe -> Just (tc, [_targ])] | tc == GHC.listTyCon ->
+                    \(cls, v, args) ->
+                        Just $ compileDict cls v args Dictionary.List.selectorsOrd
+                _ -> const Nothing
+            )
+          ]
+
+-- | Compile a class dictionary using alternative method definitions.
+--
+-- This is useful because some class instances' methods either don't have unfoldings,
+-- or cannot be directly compiled. Or we may want to use an alternative definition
+-- because it is more efficient.
+--
+-- For instance, the unfolding of `== @Integer` contains primitive operations that the
+-- plugin cannot handle, so we compile the dictionary of `Eq Integer` by using an
+-- alternative `==` from `PlutusTx.Compiler.Dictionary.Integer`.
+--
+-- Another example is `== [a]` which does not have unfolding. So we compile the dictionary
+-- of `Eq [a]` by using an alternative `==` from `PlutusTx.Compiler.Dictionary.List`.
+compileDict ::
+    CompilingDefault uni fun m ann =>
+    -- | The class whose dictionary we are compiling
+    GHC.Class ->
+    -- | Dictionary data constructor
+    GHC.Var ->
+    -- | Arguments to the dictionary data constructor
+    [GHC.CoreExpr] ->
+    -- | A list of dictionary method names where alternative definitions exist
+    [TH.Name] ->
+    m (PIRTerm uni fun)
+compileDict cls v args0 txSelNames = do
```

Comment:
This is all super fiddly and I don't feel very good about it. I'd have to stare at it a while to have ideas though.